api/uptime-crawler.js

```diff
@@ -20,12 +20,51 @@ function getClient() {
 const client = getClient();
 const setAsync = promisify(client.set).bind(client);
 
-// FIXME: this should be a genesis block API call (eventually), see:
-// https://github.com/solana-labs/solana/blob/master/cli/src/wallet.rs#L680
-// https://github.com/solana-labs/solana/blob/master/sdk/src/timing.rs#L14
-const SLOTS_PER_EPOCH = 8192;
+function getSlotsInEpoch(epochSchedule, epoch) {
+  if (!epochSchedule.warmup || epoch >= epochSchedule.first_normal_epoch) {
+    return epochSchedule.slots_per_epoch;
+  }
+
+  return null;
+}
+
+function getUptime(epochSchedule, voteState, lat, ts) {
+  const uptime = _.reduce(
+    voteState.epochCredits,
```

Comment:
What if a validator's voteState doesn't start at the first_normal_epoch? Won't that make their uptime much higher for the stage than it should be?
I think what you have is correct for mainnet, but for TdS we need to start at first_normal_epoch regardless of when the validator actually started voting.

Reply:
@CriesofCarrots that's a great point, I'll look at this right now to try to sort all the edge cases out (or at least as many as possible)!